pytest/tests/sanity/state_parts_dump_check.py

```diff
@@ -0,0 +1,126 @@
+    metrics = [line.split(' ')
+               for line in str(metrics).strip().split('\\n')
+               if re.search('state_sync_dump_check.*{', line)]
+    metrics = {metric: int(val) for metric, val in metrics}
+    return metrics
+
+
+# Waits until the target node reaches the target block
+def wait_until(target, target_node):
+    last_height = -1
+    while True:
+        last_block = target_node.get_latest_block()
+        height = last_block.height
+        if height > target:
+            break
+        if height != last_height:
+            logger.info(
+                f'@{height}, epoch_height: {state_sync_lib.approximate_epoch_height(height, EPOCH_LENGTH)}'
+            )
+            last_height = height
+
+
+def main():
+    node_config_dump, node_config = state_sync_lib.get_state_sync_configs_pair(
+    )
+    config = load_config()
+    near_root, node_dirs = init_cluster(1, 2, 4, config,
+                                        [["epoch_length", EPOCH_LENGTH]], {
+                                            0: node_config,
+                                            1: node_config_dump,
+                                            2: node_config
+                                        })
+
+    boot_node = spin_up_node(config, near_root, node_dirs[0], 0)
+    logger.info('Started boot_node')
+    dump_node = spin_up_node(config,
+                             near_root,
+                             node_dirs[1],
+                             1,
+                             boot_node=boot_node)
+    dump_check = spin_up_node(config,
+                              near_root,
+                              node_dirs[2],
+                              2,
+                              boot_node=boot_node)
+    dump_check.kill()
+    chain_id = boot_node.get_status()['chain_id']
+    dump_folder = node_config_dump["state_sync"]["dump"]["location"]["Filesystem"]["root_dir"]
+    rpc_address, rpc_port = boot_node.rpc_addr()
+    local_rpc_address, local_rpc_port = dump_check.rpc_addr()
+    logger.info(f"{rpc_address}:{rpc_port}")
+    dump_check.run_cmd(
+        cmd=(
+            'state-parts-dump-check',
+            '--chain-id',
+            chain_id,
+            '--root-dir',
+            dump_folder,
+            'loop-check',
+            '--rpc-server-addr',
+            f"http://{rpc_address}:{rpc_port}",
+            '--prometheus-addr',
+            f"{local_rpc_address}:{local_rpc_port}",
+            '--interval',
+            '2'))
+
+    logger.info('Started nodes')
+
+    wait_until(EPOCH_LENGTH - 5, boot_node)
+    # kill dumper node so that it can't dump state.
+    node_height = dump_node.get_latest_block().height
+    logger.info(f'Dump_node@{node_height}')
+    dump_node.kill()
+    logger.info(f'Killed dump_node')
+
+    wait_until(EPOCH_LENGTH + 5, boot_node)
+    # Check the dumped stats.
+    metrics = get_dump_check_metrics(dump_check)
+    assert sum([val for metric, val in metrics.items()
```

Comment:
I get an error:

```
Traceback (most recent call last):
  File "/home/marcelo/nearcore/pytest/tests/sanity/state_parts_dump_check.py", line 126, in <module>
    main()
  File "/home/marcelo/nearcore/pytest/tests/sanity/state_parts_dump_check.py", line 103, in main
    assert sum([val for metric, val in metrics.items()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

On this note, I think it would be helpful to log the object that was being checked in the assertion (and the other ones below) so that error message will show you what went wrong. e.g.: https://github.com/near/nearcore/blob/8383deb11a3089c17a1715e71a8af6215ff7c1ac/pytest/tests/sanity/skip_epoch.py#L116

But as far as this actual failure, looks like there are 4 of these with value `1`, one for each shard, so the sum is 4, not 1.

Reply:
Fixed. Added a NUM_SHARDS const to check agains.